File: A01 Array/LC128 longest Consecutive Sequence/Leetcode.py
###########
Time Complexity: O(n)
Space Complexity: O(n)


#############

class Solution:
    def longestConsecutive(self, nums):
        set_nums, ans = set(nums), 0
        for num in nums:
            if num - 1 in set_nums: continue
                
            nxt = num
            while nxt + 1 in set_nums:
                nxt += 1
            ans = max(ans, nxt-num+1)
                
        return ans
++++++++++++++++++
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        numSet= set(nums)
        longest = 0 
        
        for i in numSet:
            if (i -1) is not numSet: 
                length = 0 
                while(i + length ) in numSet:
                    length +=1
                longest = max(longest, length)
        return longest

File: A03 Graph/005 DFS connected Grid/329. Longest Increasing Path in a Matrix/Leetcode.java
/*
Time Complexity: O(MN)
Space Complexity:O(Mn)

*/ 


private int m;
private int n;
private int[][] maxILen;
private int[][] matrix;

public int longestIncreasingPath(int[][] matrix) {
    
    m = matrix.length;
    if(m < 1) return 0;
    n = matrix[0].length;
    this.matrix = matrix;
    
    maxILen = new int[m][n];
    int max = 1;
    
    for(int i=0;i<m;i++){
        
        for(int j=0;j<n;j++){
            max = Math.max(max, maxDFS(i,j));                        
        }
    }
    
    return max;
}

private int maxDFS(int i,int j){
    if(maxILen[i][j] != 0) return maxILen[i][j];
    
    int max = 1;
    if(i+1 < m && matrix[i+1][j] > matrix[i][j]) max = Math.max(max, 1 + maxDFS(i+1,j));
    if(i-1 >= 0 && matrix[i-1][j] > matrix[i][j]) max = Math.max(max, 1 + maxDFS(i-1,j));
    if(j+1 < n && matrix[i][j+1] > matrix[i][j]) max = Math.max(max, 1 + maxDFS(i,j+1));
    if(j-1 >= 0 && matrix[i][j-1] > matrix[i][j]) max = Math.max(max, 1 + maxDFS(i,j-1));
    maxILen[i][j] = max;
    return max;
}